### DISCUSSION TITLE
Questions about a project based on this

### DISCUSSION THREAD

**Discussion by DieCola85** [doudar/SmartSpin2k#400] [Ideas]
Hello guys, first of all congratulations for your project, is awesome and very well built. As a fan of DIY route, I've built few years ago an homemade trainer, which it works well. Basically is an old bike fixed on a steel frame, and the rear wheel is free to rotate on steel pipes/bearings. I also use a pair of Garmin Vector 3 pedals (power meter plus cadence) and a Garmin forerunner watch for heart rate, able to broadcast to pc. I usually use Zwift and rouvy, and it works just fine, but I would love to make it smart by using some magnets on the rear wheel, guided by a trapezoidal screw and relative nut, all connected to a servo like a 3D printer system. What I would like to do is to use your beautiful project as a base, and connect the servo to the screw directly. I've read all the wiki already and seems to be doable, but I have one question about the cadence, which is, in your project, taken from the spin bike, which I don't have in mine: is it possible to use the cadence channel being transmitted from the Garmin Vector? Or I need a different device (which I have already)  to be read as a "spin bike"? 
Thank you in advance,
Have a nice day.
Diego

**Comment by doudar** [comment]
First of all, as you go along, Post pictures! We'd love to see the progress as you go - it sounds like a really fun project.

You're all set - both power and cadence can come from the Garmin pedals. 

Let us know if we can help further!

**Comment by DieCola85** [comment]
Thanks for the reply, yes i will post the progress for sure! I will design all stuff myself since I'm a mech designer in real life (I use Siemens Solid Edge) plus I have two 3D printers so I will design and create all in house 🤟 I will use PETG filament. Thanks again!

**Reply by eMadman** [reply]
This sounds amazing!  I'd love to see how this goes!  Very curious how you'll manage rear wheel speed and how much braking force you'll be able to apply with the magnets

**Reply by DieCola85** [reply]
i'm curious as well! :) i think if the magnets will not work i will switch to classic old school brake pads. Let's see how it goes. Actually my DIY trainer is not that beautiful but quite sturdy, i will show to you in the next days :)

**Reply by doudar** [reply]
I suspect you’ll do just fine with the magnets, just make the arm that holds them a little longer than needed so you can put a few more on if the resistance isn’t high enough. Also, don’t attempt to put them on a regular brake caliper device. One side will just overcome the other and stick itself to the wheel. Ask me how I know 😅

**Reply by DieCola85** [reply]
Hi Doudar! yeah, i will make a specific holder with cells for magnets, closed, so i can keep them inside and separated. Also good hint for the arm, i think i will do it with two cells more each side. Actually i'm thinking to use 10 neodimium magnets 20x20x10 each side. 
BTW i will use a totally dedicated system, i'm drawing it in these days, and then i will print a "beta" version for testing the dimensions and position of the objects.

**Comment by DieCola85** [comment]
Hi guys, just an update: i'm about to finish the project phase, and i'm about to start to buy the stuff i need for the project. I've seen the possibility to buy the PCB board assembled and ready to be mount in the 3D printed case, but i'm struggling to know the price of that. I'm trying to buy on PCBway site but i don't know the options i need to buy it. I'm also italian so i don't know about the shipping fees. Any rough idea about this? when i finish to draw the pieces and i have the project done i will post some screenshots :)

**Reply by doudar** [reply]
I think I can get one built up and sent to you. Send a message to smartspin2k@gmail.com

**Reply by DieCola85** [reply]
hi, i'll do it right now, thank you very much :)

**Comment by DieCola85** [comment]
Time to show something: first a screenshot of my Frankenstein trainer, born from scrap stuff i had in my garage during the first lockdown Covid here in italy. Then you see two preview images of the system i've designed in past weeks. Let's see how it goes from here :D dunno if will work but..oh well, i try. Now i'm buying all stuff i need. I will need some help in the code btw, i explain the reason: i need to introduce a 0% position (when the system goes when switched off) and the 100% strenght position (closer as possible to the wheel). I need these things if i don't wanna risk to break all. I was thinking about a "zero" point, with the system very open, and then, starting from here, a 100% point (being calculated by motor revolutions number and pitch of the thread) where the system stop to move ahead. Not sure if clear. Here i need your help :D hope you like it btw.
![1](https://user-images.githubusercontent.com/84659706/204632183-a59d7d75-e467-4f7a-a2b3-04f464900f06.jpg)
[](url)
![KS®141613468 4](https://user-images.githubusercontent.com/84659706/204632189-83656602-c8d1-4530-9029-800a39705e3a.png)
![KS®1813030016 3](https://user-images.githubusercontent.com/84659706/204632200-68171c68-892f-4d24-acad-7e314da5c253.png)

**Reply by doudar** [reply]
That looks pretty cool! It should be easy to hard code your own custom end stops into the firmware. As far as releasing tension I think you could do something like. 

`if(CAD == 0){
targetPosition = FULLY_OPEN; 
}`

On this setup, it would probably be smart to install (at least one) microswitches on the end stops and attach them to unused GPIO pins on the esp32.  

How do you plan to plan to maintain lateral position of the rear wheel so it stays centered between the eddy current brakes?

**Reply by DieCola85** [reply]
Good idea about releasing tension, basically if cadence goes to zero then the position goes to open. I need to study the type and the position of the microswitches, I'm not an expert of electronics so if you have a hint for this would be appreciated 😄 maybe one for the closest position and one for the farthest? I was thinking about using something into software, basically calculating the travel distance from a "zero" point since I have the thread pitch (3mm every turn) and the revolutions of the motor. Not viable? Thanks for the hints! Very appreciated.

**Reply by DieCola85** [reply]
About the lateral position of the bike, I will add an anchor point attached to the bike frame. I will loose some of natural feel but..I can't do different, at least for now. I have an idea to build a package being able to be mounted in the bike directly. Maybe a v2.0. Let's finish the v1.0 lol

**Reply by doudar** [reply]
> Good idea about releasing tension, basically if cadence goes to zero then the position goes to open. I need to study the type and the position of the microswitches, I'm not an expert of electronics so if you have a hint for this would be appreciated 😄 maybe one for the closest position and one for the farthest? I was thinking about using something into software, basically calculating the travel distance from a "zero" point since I have the thread pitch (3mm every turn) and the revolutions of the motor. Not viable? Thanks for the hints! Very appreciated.

You could use one micro switch. The stepper can keep track of the position from there. Any micro switch that meets your mechanical requirements will work.

**Reply by DieCola85** [reply]
So maybe wiring one switch for the "zero" position and then using the stepper motor tracking for the closest position? Could be a solution? Tomorrow I will look on microswitches world, let's see if I find something interesting 😉 thanks for the help.

**Reply by DieCola85** [reply]
> You could use one micro switch. The stepper can keep track of the position from there. Any micro switch that meets your mechanical requirements will work.

This morning i was looking for some microswitches, i've found this one which is interesting.
![1](https://user-images.githubusercontent.com/84659706/204728377-ce22aaeb-7f4c-48cd-9d17-6f0ba97dfb3c.png)
I would put it at the end of the travel, on the "farthest" position, so fully open. Can you confirm that, knowing this point, using the software/stepper motor we can calculate the closest position? if yes then i will buy it and move on :) thanks again for the help and sorry for the thousands of questions, but i'm an electronic noob.

**Reply by doudar** [reply]
Yep. One at the furthest open position would work.

**Reply by DieCola85** [reply]
> Yep. One at the furthest open position would work.

Thanks Doudar! :) 
This day i've thought more about the microswitch position, and then i think the best position to place it is at the closest position, since i'm worried about the possibility to break/crash things if these magnets get too much closer, so...i will move it at the center of the system, i will draw it and then i'll show you the position to you guys.
About the type of microswitch, i think i will go with one much more simple and smaller, like this one. 
![microswitch-a-leva-pulsante-fine-corsa-micro-switch-pcb](https://user-images.githubusercontent.com/84659706/204862780-a5c2c840-1229-4857-b035-f6fb4638dd0f.jpg)
Nothing fancy but effective still :)
Thanks for the help, i very appreciate it. If you need some help regarding design/modelling i would love to give you a hand as well, just let me know.
Now i'm buying all stuff i need and then i'm ready to build the system, can't wait!

**Comment by doudar** [comment]
By the way, one thing I've thought about on a device such as this is that you could make your own power meter by adding a strain gauge here: 
![image](https://user-images.githubusercontent.com/17362216/204647489-83e2541b-cf20-45c9-9964-a6df26632329.png)

**Reply by DieCola85** [reply]
Good idea, never thought about this because I use my Garmin Vector pedals, so I get all the infos I need from there. Plus, i don't know the amount of deflection you get there, I have to build the first prototype and then study it! Actually I've bought 24 pieces of 20x20x10mm N50 neodymium magnets, so I think I have quite a lot of power to test.

**Comment by DieCola85** [comment]
Thanks @doudar for the board, the first piece of the project! :) 
![IMG_20221202_194738](https://user-images.githubusercontent.com/84659706/205369332-87cb7cab-da4a-4266-af9c-deee5ddc2114.jpg)

**Reply by doudar** [reply]
Great! I included a spare pcb with the recom which is the hardest to get part in case you need to build a new pcb for some reason.

**Reply by DieCola85** [reply]
yeah i've seen that, thank you very much :)

**Comment by DieCola85** [comment]
Hello guys, just a little update: i've finally bought all the stuff i need, and then i'm ready to start to print some stuff. Sadly i will be away home for around 2 weeks, but at the start of January i will be full on this project. Also i've added a microswitch as hard end for the closest position, so it should be able to stop the system at a safe position, preventing a crash with wheel. Have a nice day! :) 
![wip](https://user-images.githubusercontent.com/84659706/207663335-4cfb6ffe-42f6-4e62-a7e2-7d59be048924.png)

**Comment by DieCola85** [comment]
Hi all, and happy new year everyone :)
just a quick update, project is not dead, at the opposite, i'm about to finish printing my prototype, i hope next week i will end to make the pieces i need, then i will start to mount it :) I hope it works, but very fun project so far.

**Comment by DieCola85** [comment]
Hi guys, first mockup is done :) some things to improve but looks good so far. Soon i will need some help with software :D hope you like it.
![IMG_20230118_181601](https://user-images.githubusercontent.com/84659706/213253967-e228842d-0d75-4d83-b914-cdb6f456540a.jpg)
![IMG_20230118_181552](https://user-images.githubusercontent.com/84659706/213253981-1e30cb00-70d6-4e05-bc6f-0e4b0737802b.jpg)
![IMG_20230118_181539](https://user-images.githubusercontent.com/84659706/213253994-252767ec-fd1b-4b03-a677-5faa223c61c9.jpg)
Thanks for the help and support so far.

**Reply by doudar** [reply]
That's impressive! Looks like it's going great!

**Reply by DieCola85** [reply]
thanks @doudar happy you like it! in next weeks i hope to see it moving :D

**Comment by DieCola85** [comment]
Hi friends, sorry for the lack of news, but project is not dead, at the opposite. Basically i had to design a system which "blocks" the traslation of the bike along the rollers, since, with this system, you need to have the bike firmly blocked on the middle of the system. Basically, without this, the bike can "drift" a bit on the sides of the trainer, and then "touch" the plates with the magnets, which is not good. Actually i'm working on this, i hope in the next weeks to show you finally some things in movement :)

**Comment by DieCola85** [comment]
Also, one question: is the software still compatible with V2 version of the motherboard? because i don't have the V3 version of it, i've bought from you the V2, but i'm looking for a confirm that the Motherboard V2 is still usable with the current software. 
Thank you in advance, have a nice day!

**Comment by doudar** [comment]
> Also, one question: is the software still compatible with V2 version of the motherboard? because i don't have the V3 version of it, i've bought from you the V2, but i'm looking for a confirm that the Motherboard V2 is still usable with the current software. 
> 
> Thank you in advance, have a nice day!

Yep! The firmware is completely reverse compatible.

**Reply by DieCola85** [reply]
Nice, thank you very much sir :)

**Comment by eMadman** [comment]
I came across this on my youtube feed.  There's some fantastic information here that crosses over into what you're doing here
https://www.youtube.com/watch?v=Dg8oVR4k5Dk

**Comment by DieCola85** [comment]
Hi! thanks for the link, very appreciated :)

**Comment by DieCola85** [comment]
Hi guys, what's up here? just a little update about this little fun project. I'm fighting against the screw "wobbling" effect (quite common in the 3D printing world); basically the screw "wobbles" itself and the entire system gets stuck because the motor doesnt have enough power to "win" the error. I'm trying to design some kind of joint between the screw and the motor, without success. I'm thinking about using a different engine (the nema 17 we are yousing has 40Ncm of torque), but the limit is the board driver which is 2.0 A. I'm thinking to use the biggest Nema 17 motor, which should have 65Ncm of torque..but i want to ask you guys, do you know a much better engine that stays under the 2.0 A limit? actually i'm quite stuck sadly. Have a nice evening, and thanks in advance.

**Reply by doudar** [reply]
If you're using rev 2, I'd just break out (attach wires to) the en, step and Dir connection off the board and hook it up to a larger stepper driver that suits your current needs. Something like a DM542T would work.

**Reply by DieCola85** [reply]
Thanks man, i'm not an expert, but i will try to send your reply to my friends that are more into electronics. I think, for the first step, i would try the biggest Nema 17 motor available. If not then i think i will go to the stepper driver+bigger engine route. I will let you know! :D

**Reply by doudar** [reply]
Also, make sure the pot on your stepper is turned all the way to max and the current is set to max in the firmware as well

**Reply by DieCola85** [reply]
yeah, i will look into it, thanks!

**Reply by DieCola85** [reply]
> Also, make sure the pot on your stepper is turned all the way to max

Hi Doudar, can you explain more this point? is it possible to tune it directly? or do you mean by using some specific driver/command?
thank you in advance!

**Reply by eMadman** [reply]
> > Also, make sure the pot on your stepper is turned all the way to max
> 
> Hi Doudar, can you explain more this point? is it possible to tune it directly? or do you mean by using some specific driver/command? thank you in advance!

There's a little potentiometer for adjusting vref on the TMC stepsticks.  Turn it clockwise with a precision screwdriver.  Be gentle with it. 
![image](https://user-images.githubusercontent.com/24726844/235637483-b16487d2-9fd9-4db2-a328-858a777563f1.png)

Side note, I think this might be a drop-in replacement for TMC2209 stepsticks that can take up to 48v, 6a - popular for high speed 3d printers.  Just do some research on it - i'm not 100% sure on it.  https://www.aliexpress.com/item/1005004229433552.html

**Reply by DieCola85** [reply]
nice, thanks! i will let you know :)

**Comment by DieCola85** [comment]
Hi guys, just a little update and question: i'm successfully completed my prototype (there are some flaws still, need to study the solution), btw i'm trying to configure my bluetooth devices using the "scanner" feature. 
I'm using as powermeter and cadence sensors my garmin vector 3, and as heartrate monitor, my garmin watch (forerunner 735XT). By using the bluetooth scanner, the software itself finds the "BLExxx" device which is actually the garmin vector (in fact finds two garmin vector identical devices, maybe both pedals? dunno), and doesn't find my watch as heartrate monitor. So, under the "power meter" dropdown i use one of two BLE devices (the garmin vector pedals) and as heartrate monitor i leave "any". Also there's a third voice/dropdown (which is not showed in your Wiki) and i leave it as "none". 
So what's the issue? when i try to use Zwift, at device setup, it finds under "power meter" the Smartspin device, but can't select "controllable", and also doesn't find the cadence sensor and the heartrate device as showed in your Wiki. 
Hope is clear, i think i need some help here :D thank you in advance!!

**Reply by doudar** [reply]
Connect as a power meter, then some seconds later, controllable trainer will be able to be selected.

Your watch probably doesn't support BLE heart rate. 

I haven't used the Garmin vector pedals, but you probably need to connect to the left pedal. If power is only 1/2 of normal, you can then either change it in the garmin settings or use double power in the SS2K settings. 

The last setting is for using a BLE remote as a shifter button which is beta and should be left to "none".

**Reply by DieCola85** [reply]
thank you very much for the advices, tomorrow i will fudge a bit more and let you know for sure! 
btw, for non-expert programmer like me, it took a while to understand that http://smartspin2k.local/ doesn't work like it is (i mean, maybe just for me) but had to connect it directly by using the ip address. 
Stay tuned for some updates!

**Comment by DieCola85** [comment]
It seems i have no luck, but i've done a little step ahead: i've got my smartspin2k working with my garmin pedals, meaning that at **/status** page i see wattage and cadence reading correctly, BUT, in zwift setup page, when i select the smartspin2k as powermeter device, i just get "no signal" ,and then under "controllable" device there's anything. Any idea to solve this? thank you in advance.

**Reply by eMadman** [reply]
Being a V2 PCB... it's possible you have a faulty ESP32.  This was a common issue I had building the v2's.  I was throwing away about 30% of the esp32's I bought.  Sometimes entire batches of ~10 units would arrive faulty.  Try doing a firmware reset in the /settings.html page.  You may want to try flashing the firmware all over again as well.

I'm also curious if there's any metal causing interference between the esp32 and the device you're running zwift on.  

While testing, it's also worth trying on a different phone/computer - sometimes the crappy ones would be less crappy on a different device.

If it turns out to be hardware, email me at emadmanmakes@gmail.com and I'll get you on a V3.  Downside with the newer one is that the stepper motor driver is embedded on the PCB so you won't be able to swap over to a high voltage option like the TMC5160

**Reply by DieCola85** [reply]
I will do more tests tomorrow and let you know, thanks for the reply as always.

**Reply by DieCola85** [reply]
Hi @eMadman , this morning i've tried all of your advices and it doesn't work still, i get "no signal" in zwift. I will send you an email. I'm sorry about this.

**Reply by DieCola85** [reply]
thanks for your support @eMadman ! the V3 board is working, and also my system finally it *kind of* working. The magnets system works, i need to tweak a bit the entire system, but in the next weeks i will work on it for sure, and then when it will be ready i will show you something :) thanks again for the support, very appreciated.

**Reply by eMadman** [reply]
That's amazing!  I'm glad you're getting the torque you need out of the v3 for this application as well

**Reply by DieCola85** [reply]
yeah it works fine! i think i'm maybe too much "strong" for a magnet system, but i want to try some tweaks regarding the motion side.

**Reply by doudar** [reply]
any updates @DieCola85 ?

**Reply by DieCola85** [reply]
hi doudar! in these months i was quite busy with some other projects in my life, but in the next weeks i would like to test something more on this project. Basically, i think my screw has too much play (pitch was too much big), and didn't get any fluid movement. Magnets work, system is "kind of" working, but i would like to test a "classic" metric thread and relative nut. M10 has half of the pitch of the thread i actually have, so i should be able to get more precision/small steps. I have to rebuild some parts and so on. So, i want to update my project, i just need some freetime to do it ahah 👍 i'll post some update when ready. Thanks for the interest btw!
